src/eg/src/verifiable_decryption.rs

```diff
@@ -0,0 +1,762 @@
+            let c_i = (c.clone() * w_i) % q;
+            c_i_vec.push(c_i);
+        }
+
+        // Check Equations (74) and (75)
+        for (ds, cs, rs, c_i) in izip!(
+            decryption_shares,
+            proof_commit_shares,
+            proof_response_shares,
+            c_i_vec
+        ) {
+            let g_v = g.modpow(&rs.v_i, p);
+            let i = BigUint::from(ds.i.get_one_based_u32());
+            let k_prod = guardian_public_keys
+                .iter()
+                .fold(BigUint::one(), |prod, pk| {
+                    let inner_p = pk.coefficient_commitments.0.iter().enumerate().fold(
+                        BigUint::one(),
+                        |prod, (m, k_m)| {
+                            //This is fine as m < k
+                            #[allow(clippy::unwrap_used)]
+                            let m: u32 = m.try_into().unwrap();
+                            (prod * k_m.0.modpow(&i.pow(m), p)) % p
+                        },
+                    );
+                    (prod * inner_p) % p
+                });
+            let a_i = (g_v * k_prod.modpow(&c_i, p)) % p;
+
+            let a_v = ciphertext.alpha.modpow(&rs.v_i, p);
+            let m_c = ds.m_i.modpow(&c_i, p);
+            let b_i = (a_v * m_c) % p;
+
+            if a_i != cs.a_i {
+                return Err(CombineProofError::CommitInconsistency(
+                    ds.i,
+                    "a_i != a_i'".into(),
+                ));
+            }
+            if b_i != cs.b_i {
+                return Err(CombineProofError::CommitInconsistency(
+                    ds.i,
+                    "b_i != b_i'".into(),
+                ));
+            }
+        }
+
+        let v = proof_response_shares
+            .iter()
+            .fold(BigUint::zero(), |sum, rs| (sum + rs.v_i.clone()) % q);
+
+        Ok(DecryptionProof {
+            challenge: c,
+            response: v,
+        })
+    }
+
+    /// This function validates a decryption proof.
+    ///
+    /// The arguments are
+    /// - `self` - the decryption proof
+    /// - `fixed_parameters` - the fixed parameters
+    /// - `h_e` - the extended bash hash
+    /// - `joint_key` - the joint election public key
+    /// - `ciphertext` - the ciphertext
+    /// - `m` - combined decryption share
+    pub fn validate(
+        &self,
+        fixed_parameters: &FixedParameters,
+        h_e: &HValue,
+        joint_key: &JointElectionPublicKey,
+        ciphertext: &Ciphertext,
+        m: &CombinedDecryptionShare,
+    ) -> bool {
+        let g = &fixed_parameters.g;
+        let p = fixed_parameters.p.as_ref();
+
+        let key = &joint_key.joint_election_public_key;
+
+        let a = (g.modpow(&self.response, p) * key.modpow(&self.challenge, p)) % p;
+        let b = (ciphertext.alpha.modpow(&self.response, p) * m.0.modpow(&self.challenge, p)) % p;
+
+        //Check (9.A)
+        if &self.response >= p {
```

Comment:
Should this not be q according to the specification?

Reply:
Good catch! Yes it should be `q`. I have fixed it.